### PR TITLE
Queue properties

### PR DESCRIPTION
**Pull request by j3mdamas**
Uniformization of SimQueues.  …
   * Creation of (abstract) generic properties that can be probed for any implemented SimQueue: ncpus, ngpus, memory
   * Generalization of the wait method (abstract method on SimQueue now), to include sentinel files and notcompleted mode (all in SimQueue)
   * super __init__ call replaced by separate calls to each parent, as it was not working (when I created an __init__ for SimQueue)
   * LocalQueues:
        - run_job converted into method
        - _getmemory method implemented
        - ncpus of LocalQueue detached from the ncpus of htmd

 TODO:
     - Merge LocalGPUQueue and LocalCPUQueue into a single class which manages both ncpu and ngpu simultaneously
     - Implement generic properties and notcompleted on AceCloudQueue

### CODE REVIEW

File: htmd/queues/localqueue.py [hunk 1]
```diff
@@ -117,7 +154,9 @@ def submit(self, mydirs):
         """
         self._setupQueue()
 
-        if isinstance(mydirs, str): mydirs = [mydirs]
+        if isinstance(mydirs, str):
+            mydirs = [mydirs]
+        self._dirs.extend(mydirs)
```
Comment (stefdoerr) [inline htmd/queues/localqueue.py]: why here and not in .submit?

File: htmd/queues/localqueue.py [hunk 2]
```diff
@@ -144,12 +183,37 @@ def inprogress(self):
...
     def _getdevices(self):
-        pass
+        return list()
+
+    def _getmemory(self):
+        ureg = UnitRegistry()
+        total_memory = int(ureg.Quantity(psutil.virtual_memory().total, ureg.byte).to('MiB').magnitude)
+        nr_devices = len(self._getdevices())
+        if nr_devices != 0:
+            return int(total_memory/nr_devices)
```
Comment (stefdoerr) [inline htmd/queues/localqueue.py]: This is horribly nasty btw. It assumes the computer has 0 memory in use.

Comment (j3mdamas) [reply htmd/queues/localqueue.py]: virtual_memory().total is all total memory available, not only the available.

what do you propose?

File: htmd/queues/localqueue.py (hunk 1 above)
Comment (j3mdamas) [reply htmd/queues/localqueue.py]: oh right, this is a bug, I'll change it.

Comment (stefdoerr) [reply htmd/queues/localqueue.py]: ah forget it. bad diff viewer

**Comment by stefdoerr** [comment]
a) that does not tell me why they have to be properties. You could make them normal "args" of the class like the rest of the options, no?
b) yes I am actually proposing on relying on disk information but keeping tabs of simulations using a disk-based database

**Comment by j3mdamas** [comment]
a) as properties, you can oblige them to be defined in child classes, and also standardizes naming (queue vs. partition problem).
b) sure!

**Comment by stefdoerr** [comment]
Ah yes ok for the standardization. right. Otherwise it's two separate
variables
Ok I'm fine with it

**Comment by raimis** [comment]
@j3mdamas @stefdoerr 

`LocalCPUQueue` is still broken:
```python
>>> from htmd.queues.localqueue import LocalCPUQueue
>>> q = LocalCPUQueue()
>>> print(q.memory)
None
>>> print(q.ncpu)
None
```
Some CPUs can be found! 
```python
>>> q._getdevices()
[None, None, None, None, None, None, None, None]
```
Well, I expected a bit more...
```python
>>> q._getmemory()
>>> 1978
```
```bash
$ free -m
              total        used        free      shared  buff/cache   available
Mem:          15830        2961        9031         702        3837       11773
Swap:         16383           0       16383
```

**Comment by stefdoerr** [comment]
Works as expected I'd say.
ncpu is not available cpus. It's how many you have set (which you haven't)

**Comment by stefdoerr** [comment]
Same for `q.memory`

**Comment by stefdoerr** [comment]
Just look at the code really.
```python
    def _getmemory(self):
        ureg = UnitRegistry()
        total_memory = int(ureg.Quantity(psutil.virtual_memory().total, ureg.byte).to('MiB').magnitude)
        nr_devices = len(self._getdevices())
        if nr_devices != 0:
            return int(total_memory/nr_devices)
        else:
            return None
```
As you can see it divides the memory by number of devices.
I don't know if I agree with this way, I'd personally report whole memory and then divide outside the function but that's up to @j3mdamas

**Comment by j3mdamas** [comment]
Just fixed it with @raimis. I've changed the `LocalCPUQueue` to have defaults.

**Pull request merged**